crypto.go

```diff
@@ -0,0 +1,149 @@
+package sphinx
+
+import (
+	"crypto/hmac"
+	"crypto/sha256"
+
+	"github.com/aead/chacha20"
+	"github.com/btcsuite/btcd/btcec"
+)
+
+const (
+	// HMACSize is the length of the HMACs used to verify the integrity of
+	// the onion. Any value lower than 32 will truncate the HMAC both
+	// during onion creation as well as during the verification.
+	HMACSize = 32
+)
+
+// Hash256 is a statically sized, 32-byte array, typically containing
+// the output of a SHA256 hash.
+type Hash256 [sha256.Size]byte
+
+// zeroHMAC is the special HMAC value that allows the final node to determine
+// if it is the payment destination or not.
+var zeroHMAC [HMACSize]byte
+
+// calcMac calculates HMAC-SHA-256 over the message using the passed secret key
+// as input to the HMAC.
+func calcMac(key [keyLen]byte, msg []byte) [HMACSize]byte {
+	hmac := hmac.New(sha256.New, key[:])
+	hmac.Write(msg)
+	h := hmac.Sum(nil)
+
+	var mac [HMACSize]byte
+	copy(mac[:], h[:HMACSize])
+
+	return mac
+}
+
+// xor computes the byte wise XOR of a and b, storing the result in dst.
```

Comment:
should comment note that only the first `min(len(a), len(b)` bytes will be xor'd?